tests/components/tcp/test_sensor.py

```diff
@@ -36,228 +37,232 @@
+            assert await async_setup_component(hass, "sensor", {"tcp": config})
+
+
+def test_init_calls_update(hass):
+    """Call update() method during __init__()."""
+    with patch("homeassistant.components.tcp.sensor.TcpSensor.update") as mock_update:
+        tcp.TcpSensor(hass, TEST_CONFIG)
+        assert mock_update.called
+
+
+def test_update_connects_to_host_and_port(hass, mock_select, mock_socket):
+    """Connect to the configured host and port."""
+    tcp.TcpSensor(hass, TEST_CONFIG["sensor"])
+    mock_socket = mock_socket().__enter__()
+    assert mock_socket.connect.mock_calls[0][1] == (
+        (
+            TEST_CONFIG["sensor"][tcp.CONF_HOST],
+            TEST_CONFIG["sensor"][tcp.CONF_PORT],
+        ),
+    )
+
+
+def test_update_returns_if_connecting_fails(hass, *args):
+    """Return if connecting to host fails."""
+    with patch("homeassistant.components.tcp.sensor.TcpSensor.update"):
+        sensor = tcp.TcpSensor(hass, TEST_CONFIG["sensor"])
+    assert sensor.update() is None
+
+
+@patch("socket.socket.send", side_effect=socket.error())
+def test_update_returns_if_sending_fails(hass, *args):
+    """Return if sending fails."""
+    with patch("homeassistant.components.tcp.sensor.TcpSensor.update"):
+        sensor = tcp.TcpSensor(hass, TEST_CONFIG["sensor"])
+    assert sensor.update() is None
+
+
+@patch("socket.socket.send")
+def test_update_returns_if_select_fails(hass, *args):
+    """Return if select fails to return a socket."""
+    with patch("homeassistant.components.tcp.sensor.TcpSensor.update"):
+        sensor = tcp.TcpSensor(hass, TEST_CONFIG["sensor"])
+    assert sensor.update() is None
+
+
+def test_update_sends_payload(hass, mock_select, mock_socket):
+    """Send the configured payload as bytes."""
+    tcp.TcpSensor(hass, TEST_CONFIG["sensor"])
+    mock_socket = mock_socket().__enter__()
+    mock_socket.send.assert_called_with(
+        TEST_CONFIG["sensor"][tcp.CONF_PAYLOAD].encode()
+    )
+
+
+def test_update_calls_select_with_timeout(hass, mock_select, mock_socket):
+    """Provide the timeout argument to select."""
+    tcp.TcpSensor(hass, TEST_CONFIG["sensor"])
+    mock_socket = mock_socket().__enter__()
+    mock_select.assert_called_with(
+        [mock_socket], [], [], TEST_CONFIG["sensor"][tcp.CONF_TIMEOUT]
+    )
+
+
+def test_update_receives_packet_and_sets_as_state(hass, mock_select, mock_socket):
+    """Test the response from the socket and set it as the state."""
+    test_value = "test_value"
+    mock_socket = mock_socket().__enter__()
+    mock_socket.recv.return_value = test_value.encode()
+    config = copy(TEST_CONFIG["sensor"])
+    del config[tcp.CONF_VALUE_TEMPLATE]
+    sensor = tcp.TcpSensor(hass, config)
+    assert sensor._state == test_value
+
+
+def test_update_renders_value_in_template(hass, mock_select, mock_socket):
+    """Render the value in the provided template."""
+    test_value = "test_value"
+    with patch("socket.socket") as mock_socket:
```

Comment:
We already mock socket. We can set the recv attribute return value here.

Reply:
Done, however still struggling with getting these tests to pass (see my first comment on this PR).